File: prototype/entities.py
class Person:
    def __init__(self, person_id: str, name: str, role: str):
        self.id = person_id
        self.name = name
        self.role = role  # "PL" / "DEV" / "CUSTOMER" / "BOSS" など

    def speak(self, current_task=None) -> str:
        """キャラクターの発言を取得する（子クラスでオーバーライド）"""
        return ""


class Developer(Person):
    def __init__(self, dev_id: str, name: str, work_speed: float, base_bug_rate: float, salary: int, personality_tags: list, role: str = "DEV", specialty: str = "BE"):
        super().__init__(dev_id, name, role)
        self.work_speed = work_speed
        self.base_bug_rate = base_bug_rate
        self.salary = salary  # 日当 (1日のコスト)
        self.personality_tags = personality_tags
        self.specialty = specialty  # "FE" / "BE"
        
        # 隠しパラメータ
        self._morale = 80.0  # 士気 (0-100)
        self._fatigue = 0.0  # 疲労 (0-100)
        
        # パラメータ一時開示期限（週数）
        self.reveal_duration = 0

    @property
    def morale(self):
        return self._morale

    @morale.setter
    def morale(self, value):
        self._morale = max(0.0, min(100.0, value))

    @property
    def fatigue(self):
        return self._fatigue

    @fatigue.setter
    def fatigue(self, value):
        self._fatigue = max(0.0, min(100.0, value))

    def speak(self, current_task=None) -> str:
        """開発者の状態に応じた『ひと言サイン』を出力する"""
        return self.get_sign(current_task)

    def get_sign(self, current_task=None) -> str:
        """開発者の状態に応じた『ひと言サイン』を出力する"""
        if self.role == "PL":
            if self.fatigue >= 80 or self.morale <= 20:
                return "「メンバーも疲弊してますし、私ももう限界です……。進捗管理どころではありません。」"
            elif self.fatigue >= 50 or self.morale <= 50:
                return "「PM、現場に直接口を出しすぎではないですか？私への相談を通してください。」"
            
            # PLの専門性（相性）に応じたメッセージ
            if self.specialty == "BE":  # システム内部でのマッチング
                return "「今回のプロジェクト要件は私の得意ドメインなので、設計は任せてください。PMは交渉に専念を。」"
            elif self.specialty == "FE":
                return "「今回のシステム要件は以前にも同様の経験があります。現場管理は私に任せてください。」"
            return "「進捗管理は私に任せて、PMは顧客交渉やリスク対策に集中してください。」"

        # DEVの場合
        # 1. 限界状態
        if self.fatigue >= 80 or self.morale <= 20:
            return "「……うう、頭が痛いです。体調が優れないので作業が遅れるかもしれません……」"
            
        # 2. ミスマッチ状態（現在作業中のタスクがある場合）
        if current_task and current_task.skill_type != self.specialty:
            if self.fatigue >= 50 or self.morale <= 50:
                return "「経験の薄い分野のタスクで、しかも疲れが溜まっていて全然頭が回りません……」"
            return "「今回の案件の要件はあまり経験がない分野なんですよね……少し手探りです」"
            
        # 3. 疲労蓄積（要注意）
        if self.fatigue >= 50 or self.morale <= 50:
            return "「最近ちょっと寝不足ですね……。仕様がコロコロ変わると心身ともにキツいです」"
        
        # 通常・良好な時
        if "DRINK_LOVER" in self.personality_tags and self.morale >= 85:
            return "担当タスクを笑顔でこなしています。「今度みんなで飲みに行きませんか？」"
        if "TECH_GEEK" in self.personality_tags and self.morale >= 85:
            return "黙々と作業しています。「新しい設計フレームを導入したら、品質が上がりました！」"
        
        return "「今週も順調です！タスクを進めていきます」"


class Task:
    def __init__(self, task_id: str, name: str, estimated_hours: float, skill_type: str = "BE"):
        self.id = task_id
        self.name = name
        self.estimated_hours = estimated_hours
        self.actual_hours = 0.0
        self.progress = 0.0  # 0.0 - 100.0
        self.assigned_developer_id = None
        self.status = "TODO"  # "TODO", "IN_PROGRESS", "DONE"
        self.skill_type = skill_type  # "FE" / "BE"


class Customer(Person):
    def __init__(self, customer_id: str, name: str, customer_type: str):
        super().__init__(customer_id, name, "CUSTOMER")
        self.type = customer_type  # "SPEED_ORIENTED", "QUALITY_ORIENTED", "VAGUE_REQUIREMENTS"
        self.satisfaction = 80.0  # 0 - 100
        self.vague_level = 80.0   # あいまい度 (0 - 100)

    def speak(self, current_task=None) -> str:
        return f"「私は{self.name}です。タイプは{self.type}です。」"


class Project:
    def __init__(self, name: str, budget: int, deadline_weeks: int, customer: Customer, clarity_level: int = 3, budget_level: int = 3, schedule_level: int = 3):
        self.name = name
        self.budget = budget
        self.deadline_weeks = deadline_weeks
        self.customer = customer
        
        # 3つのレベル感パラメータ (1〜5)
        self.clarity_level = clarity_level
        self.budget_level = budget_level
        self.schedule_level = schedule_level
        
        self.bugs_total = 0
        self.reported_bugs = 0  # 顧客・上司に報告済みのバグ数
        
        self.manager_satisfaction = 80.0  # 上級マネージャー満足度 (0 - 100)
        self.week = 1
        
        # 交渉による約束・猶予
        self.deadline_extension_weeks = 0  # 顧客に認められた延長週数
        self.extra_budget = 0             # 上司から獲得した追加予算

        # PL関連ステータス
        self.pl_active = True       # PLが自律稼働しているか
        self.direction = "NORMAL"   # 現在の開発方針 ("NORMAL" / "BUG_FIRST")
        
        # 雇用中のメンバーリスト (ゲーム中の体制)
        self.assigned_developers = []  # List of Developer

        # プロジェクト立ち上げ・事前準備状態管理用
        self.hearing_type = None    # ヒアリングのタイプ ("DEEP" / "LIGHT" / "NONE")
        self.has_evidence = False   # エビデンス（見積もりレポート確認済）の有無

File: web/src/entities.js
// Web Prototype Data Models (Entities)

export class Developer {
  constructor(id, name, options = {}) {
    this.id = id;
    this.name = name;
    this.age = options.age || 25;
    this.techSkill = options.techSkill || 3;
    this.commSkill = options.commSkill || 3;
    this.leadershipSkill = options.leadershipSkill || 2;
    this.speedSkill = options.speedSkill || 3;
    this.mentalSkill = options.mentalSkill || 3;
    
    this.fatigue = options.fatigue || 0.0;
    this.morale = options.morale || 80.0;
    this.resolution = options.resolution || 0; // 0: 未知, 1: シルエット, 2: 開示
    this.assignedRole = "DEV"; // "PL" or "DEV"
    this.isRetired = false;
  }

  get isPlQualified() {
    return this.leadershipSkill >= 3;
  }

  get statusDisplay() {
    if (this.resolution === 0) {
      return `未知 (?) | 年齢: ${this.age}歳`;
    } else if (this.resolution === 1) {
      const fLabel = this.fatigue >= 70 ? "高" : (this.fatigue >= 40 ? "中" : "良好");
      return `概算理解 | 疲労感: ${fLabel} | 年齢: ${this.age}歳`;
    } else {
      return `詳細開示 | 疲労: ${this.fatigue.toFixed(0)}/100 | 士気: ${this.morale.toFixed(0)}/100 | 統率: ${this.leadershipSkill}/5`;
    }
  }
}

export class Customer {
  constructor(id, name, type = "QUALITY_ORIENTED") {
    this.id = id;
    this.name = name;
    this.type = type; // QUALITY_ORIENTED, SPEED_ORIENTED, VAGUE_REQUIREMENTS
    this.satisfaction = 70.0;
    this.revealed = false;
  }

  speak() {
    if (this.type === "QUALITY_ORIENTED") {
      return "「障害やバグは絶対に許されません。品質第一でお願いしますよ。」";
    } else if (this.type === "SPEED_ORIENTED") {
      return "「とにかくスピード重視です！納期遅れは致命的ですよ。」";
    } else {
      return "「要件はまだ固まっていません。柔軟に対応してくださいね。」";
    }
  }
}

export class Task {
  constructor(id, name, estimatedHours) {
    this.id = id;
    this.name = name;
    this.estimatedHours = estimatedHours;
    this.progress = 0.0; // 0 - 100
    this.status = "TODO"; // TODO, IN_PROGRESS, DONE
    this.assignedDeveloperId = null;
  }
}

export class Team {
  constructor(id, name) {
    this.id = id;
    this.name = name;
    this.leader = null;
    this.members = [];
  }

  setLeader(developer) {
    this.leader = developer;
    developer.assignedRole = "PL";
  }

  assignMember(developer) {
    this.members.push(developer);
    developer.assignedRole = "DEV";
  }

  get allMembers() {
    const list = [];
    if (this.leader) list.push(this.leader);
    return list.concat(this.members);
  }
}

export class Project {
  constructor(name, deadlineWeeks, customer) {
    this.name = name;
    this.deadlineWeeks = deadlineWeeks;
    this.customer = customer || new Customer("cust_1", "顧客部長");
    this.week = 1;
    this.day = 1;
    this.maxDays = deadlineWeeks * 5; // 4週間 = 20営業日
    this.scheduledMeetings = []; // [{ day: 5, actionId: "req_def_ws", title: "要件定義WS" }]
    this.methodology = "WATERFALL"; // WATERFALL vs AGILE
    this.teams = [];
    
    // プロジェクトパラメータ
    this.clarityLevel = 3;  // 1-5
    this.budgetLevel = 3;   // 1-5
    this.scheduleLevel = 3; // 1-5
    
    this.managerSatisfaction = 70.0;
    this.direction = "NORMAL"; // NORMAL, SPEED, QUALITY
    this.bugsTotal = 0;
    this.reportedBugs = 0;
    this.priorityExpectation = "QUALITY";
  }

  // 11月2日(月)を開始基準としたリアルな月日・曜日文字列を取得ヘルパー
  getDateString(dayNum = this.day) {
    const weekdays = ["日", "月", "火", "水", "木", "金", "土"];
    // 基準日: 2026年11月2日(月)
    const current = new Date(2026, 10, 2);
    let workDaysCount = 1;
    while (workDaysCount < dayNum) {
      current.setDate(current.getDate() + 1);
      const dayOfWeek = current.getDay();
      if (dayOfWeek !== 0 && dayOfWeek !== 6) {
        workDaysCount++;
      }
    }
    const month = current.getMonth() + 1;
    const date = current.getDate();
    const wday = weekdays[current.getDay()];
    return `${month}月${date}日(${wday})`;
  }

  registerTeam(team) {
    this.teams.push(team);
  }

  get mainTeam() {
    return this.teams[0] || null;
  }

  getAllDevelopers() {
    if (!this.mainTeam) return [];
    return this.mainTeam.allMembers;
  }
}

export class PM {
  constructor(maxAp = 3) {
    this.maxAp = maxAp;
    this.ap = maxAp;
    this.careerYears = 1;
    this.completedProjects = 0;
  }

  resetAp() {
    this.ap = this.maxAp;
  }
}

// UI 遷移モード定義
export const UIMode = {
  TITLE: "TITLE",
  PROLOGUE_INTRO: "PROLOGUE_INTRO",
  PROLOGUE: "PROLOGUE",
  DASHBOARD: "DASHBOARD",
  SCENE_CUSTOMER: "SCENE_CUSTOMER",
  SCENE_MANAGER: "SCENE_MANAGER",
  SCENE_TEAM: "SCENE_TEAM",
  EVENT_MODAL: "EVENT_MODAL"
};

// アクション定義 (ADV UIの各シーンでの具体アクション)
export const ADV_ACTIONS = {
  // 顧客会議室でのアクション
  CUSTOMER: [
    {
      id: "req_def_ws",
      name: "💡 要件定義WSのアポを取る",
      targetGroup: "CUSTOMER",
      costAp: 1,
      isAppointment: true,
      delayDays: 2,
      desc: "【アポ予約】2日後に要件定義ワークショップを予約設定する"
    },
    {
      id: "phased_release",
      name: "👥 段階リリース交渉のアポを取る",
      targetGroup: "CUSTOMER",
      costAp: 1,
      isAppointment: true,
      delayDays: 3,
      desc: "【アポ予約】3日後にスコープ調整の役員面談を予約設定する"
    },
    {
      id: "prototype_demo",
      name: "📱 プロトタイプを先行提示",
      targetGroup: "CUSTOMER",
      costAp: 1,
      isAppointment: false,
      delayDays: 0,
      desc: "【即時】開発初期の認識ズレをモックで即時防止する"
    },
    {
      id: "qcd_align",
      name: "🤝 QCD優先順位の合意アポを取る",
      targetGroup: "CUSTOMER",
      costAp: 1,
      isAppointment: true,
      delayDays: 1,
      desc: "【アポ予約】翌日に期待値すり合わせ面談を予約設定する"
    }
  ],
  // 上司執務室でのアクション
  MANAGER: [
    {
      id: "buffer_request",
      name: "🏢 納期バッファ直訴のアポを取る",
      targetGroup: "MANAGER",
      costAp: 1,
      isAppointment: true,
      delayDays: 1,
      desc: "【アポ予約】翌日に上司へ納期猶予直訴の面談を予約設定する"
    },
    {
      id: "helper_request",
      name: "🙋‍♂️ 助っ人要請の面談アポを取る",
      targetGroup: "MANAGER",
      costAp: 1,
      isAppointment: true,
      delayDays: 1,
      desc: "【アポ予約】翌日に助っ人エンジニア要請の面談を予約設定する"
    },
    {
      id: "boss_risk_check",
      name: "❓ 上司のリスク許容範囲確認",
      targetGroup: "MANAGER",
      costAp: 0,
      isAppointment: false,
      delayDays: 0,
      desc: "【即時/無料】ノーリスクで上司の評価ラインと期待値を確認する"
    }
  ],
  // 開発チームデスクでのアクション (現場アクションは即時中心)
  TEAM: [
    {
      id: "holiday_work_request",
      name: "🚨 現場チームに休日出勤を依頼する (土曜出社)",
      targetGroup: "TEAM",
      costAp: 1,
      isAppointment: false,
      delayDays: 0,
      desc: "【即時】開発進捗を大幅リカバリー！ ただしメンバー疲労度+25%急上昇のジレンマ"
    },
    {
      id: "team_kickoff",
      name: "🚀 チームキックオフ・役割分担の整理 (体制構築)",
      targetGroup: "TEAM",
      costAp: 1,
      isAppointment: false,
      delayDays: 0,
      desc: "【即時】チームの顔合わせと各人の役割分担を整理し健全性を向上させる"
    },
    {
      id: "tech_risk_check",
      name: "🛠️ 技術リスク・見積精査",
      targetGroup: "TEAM",
      costAp: 1,
      isAppointment: false,
      delayDays: 0,
      desc: "【即時】現場のコード・見積精度を自ら精査し交渉の根拠を作る"
    },
    {
      id: "retrospective_share",
      name: "📚 過去の失敗教訓共有",
      targetGroup: "TEAM",
      costAp: 1,
      isAppointment: false,
      delayDays: 0,
      desc: "【即時】チームの安全度を上げバグ発生率を低減する"
    },
    {
      id: "one_on_one",
      name: "❓ チームの懸念点1on1",
      targetGroup: "TEAM",
      costAp: 0,
      isAppointment: false,
      delayDays: 0,
      desc: "【即時/無料】メンバーの隠れ不安やモチベーション状態を対話で看破する"
    }
  ]
};

